pkg/internal/java/java_inject.go

```diff
@@ -0,0 +1,351 @@
+	// Get the path to OBI
+	exePath, err := os.Executable()
+	if err != nil {
+		return "", err
+	}
+
+	// Resolve any symlinks
+	exePath, err = filepath.EvalSymlinks(exePath)
+	if err != nil {
+		return "", err
+	}
+
+	// Get the directory containing OBI
+	exeDir := filepath.Dir(exePath)
+
+	// Construct the path to the file relative to OBI
+	filePath := filepath.Join(exeDir, ObiJavaAgentFileName)
+
+	return filePath, nil
+}
+
+// to be changed in tests
+var rootDirForPID func(int32) string = ebpfcommon.RootDirectoryForPID
+
+func (i *JavaInjector) copyAgent(ie *ebpf.Instrumentable) (string, error) {
+	root := rootDirForPID(ie.FileInfo.Pid)
+	tempDir, err := i.findTempDir(root, ie)
+	if err != nil {
+		return "", fmt.Errorf("error accessing temp directory: %w", err)
+	}
+
+	fullTempDir := filepath.Join(root, tempDir)
+
+	i.log.Info("found injection directory for process", "pid", ie.FileInfo.Pid, "path", fullTempDir)
+
+	agentPathHost := filepath.Join(fullTempDir, ObiJavaAgentFileName)
+
+	source, err := os.Open(i.agentPath)
+	if err != nil {
+		return "", fmt.Errorf("unable to access OBI java agent: %w", err)
+	}
+
+	defer source.Close()
+	// Create file with read permissions for owner, group, and others (0644)
+	target, err := os.OpenFile(agentPathHost, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o644)
+	if err != nil {
+		return "", fmt.Errorf("unable to create target OBI java agent: %w", err)
+	}
+	defer target.Close()
+	if _, err = target.ReadFrom(source); err != nil {
+		return "", fmt.Errorf("error writing java agent to target location: %w", err)
+	}
+
+	agentPathContainer := filepath.Join(tempDir, ObiJavaAgentFileName)
+
+	return agentPathContainer, nil
+}
+
+func returnCodeLine(line string) (bool, error) {
+	if strings.Contains(line, "return code: 0") || strings.Contains(line, "ATTACH_ACK") {
+		return true, nil
+	} else if strings.Contains(line, "return code:") {
+		return true, fmt.Errorf("error executing command for the JVM %s", line)
+	}
+
+	return false, nil
+}
+
+func (i *JavaInjector) attachJDKAgent(attacher *jvm.JAttacher, pid int32, path string) error {
+	attacher.Init()
+
+	defer func() {
+		if err := attacher.Cleanup(); err != nil {
+			slog.Warn("error on JVM attach cleanup", "error", err)
+		}
+	}()
+	out, err := attacher.Attach(int(pid), []string{"load", "instrument", "false", path}, false)
+	if err != nil {
+		i.log.Error("error executing command for the JVM", "pid", pid, "error", err)
+		return err
+	}
+
+	defer out.Close()
+
+	reader := bufio.NewReader(out)
+	buf := bytes.Buffer{}
+	for {
+		b, err := reader.ReadByte()
+		if err != nil {
+			if err == io.EOF { // hotspot terminates with EOF
+				_, err := returnCodeLine(buf.String())
+				if err != nil {
+					return err
+				}
+			}
+			return fmt.Errorf("error reading line %w", err)
+		}
+
+		buf.WriteByte(b)
+		if b == '\n' {
+			if end, err := returnCodeLine(buf.String()); end {
+				return err
+			}
+
+			buf.Reset()
+		} else if b == 0 { // j9 terminates with 0
```

Comment:
Do you need to check if the `0` is at the end of the buffer? Otherwise it seems this conditional is true for any byte that is `0`?

Reply:
This code was translated from jattach, this is how they check for end as well.